apps/labs/posts/multi-dimensional-sparse-arrays.md

```diff
@@ -0,0 +1,225 @@
+The CSR format allows for extremely fast arithmetic and vector operations, however due to how it stores the arrays, it’s more complex to extend to n-dimensions. In contrast, the COOrdinate format facilitates very fast conversion to and from CSR, and its simpler structure is more intuitive for extending to n-dimensions. Essentially, the COO format (for 1D and 2D) stores the non-zero data and the coordinates of that data using three arrays – `data`, `row`, and `col`. When extended to n-D, we instead represent it using a `data` array and a tuple of n arrays, where each array, `a_1, …, a_i, …, a_n`, stores the indices for the i-th dimension.
+
+## Why n-D Sparse Arrays?
+
+Several fields of machine learning, such as Natural Language Processing (NLP), Computer Vision, and Recommender Systems, have input data which is almost always sparse. As the need for handling sparse data grows, it becomes essential to have a robust mechanism that can address sparsity in n-dimensions.
+Currently, SciPy's sparse arrays are limited to 1D and 2D, with higher-dimensional operations relying on dense NumPy arrays, leading to significant memory and computational inefficiencies. Extending COO sparse arrays to n-dimensions will:
+
+1. **Improve Efficiency**: Avoid dense format storage for high-dimensional sparse tensors.
+2. **Enhance Performance**: Reduce memory overhead and unnecessary computations in matrix and tensor operations.
+3. **Expand Functionality**: Enable direct manipulation of n-dimensional sparse data.
+
+---
+
+## Workflow and Implementation
+
+To kick off the process of this work, the issue [#21169](https://github.com/scipy/scipy/issues/21169) was opened.
+This laid out the planned workflow, and roughly outlined the distribution of tasks across several PRs (Pull Requests) that would be submitted over time.
+
+The implementation was carried out in phases, with each phase focusing on extending a predefined set of methods to n-dimensions. These phases structured the contents of each PR.
+
+The first stage involved developing the core constructors required to initialise an n-D COO array, as well as enabling conversions between COO and dense formats. This was followed by extending basic methods such as `reshape`, `transpose`, `eliminate_zeros` etc. to n-D. Next, element-wise arithmetic operations—addition and subtraction—were implemented. At the time, this replicated the pre-existing 1D/2D sparse array implementation of element-wise operations, where we only dealt with arrays of the same shape (this behavior was later updated when we introduced broadcasting for sparse arrays, allowing element-wise operations for all broadcastable arrays to align with NumPy's handling of such methods).
+
+By the end of this stage, I had become fairly confident that the entire COO codebase could be extended to n-dimensions.
+
+I then moved on to the second stage, which involved tackling the fundamental operations in linear algebra, `matmul`, `dot`, and `tensordot`.
```

Comment:
I suggest linking to the documentation on those - just for completeness.

Reply:
Would this be to the SciPy sparse documentation (because I haven't added docstrings for `tensordot` which is a new addition to the codebase), or linking to the NumPy documentation?